--- a/panels/display/cc-display-config.c	2022-06-30 16:12:42.000000000 -0700
+++ b/panels/display/cc-display-config.c	2022-07-07 10:52:07.240024308 -0700
@@ -334,6 +334,25 @@
   return CC_DISPLAY_MONITOR_GET_CLASS (self)->set_underscanning (self, underscanning);
 }
 
+gboolean
+cc_display_monitor_supports_variable_refresh_rate (CcDisplayMonitor *self)
+{
+  return CC_DISPLAY_MONITOR_GET_CLASS (self)->supports_variable_refresh_rate (self);
+}
+
+gboolean
+cc_display_monitor_get_variable_refresh_rate (CcDisplayMonitor *self)
+{
+  return CC_DISPLAY_MONITOR_GET_CLASS (self)->get_variable_refresh_rate (self);
+}
+
+void
+cc_display_monitor_set_variable_refresh_rate (CcDisplayMonitor *self,
+                                              gboolean enabled)
+{
+  return CC_DISPLAY_MONITOR_GET_CLASS (self)->set_variable_refresh_rate (self, enabled);
+}
+
 void
 cc_display_monitor_set_mode (CcDisplayMonitor *self, CcDisplayMode *m)
 {
--- a/panels/display/cc-display-config-dbus.c	2022-06-30 16:12:42.000000000 -0700
+++ b/panels/display/cc-display-config-dbus.c	2022-07-07 10:51:45.135601214 -0700
@@ -317,7 +317,6 @@
   gobject_class->finalize = cc_display_logical_monitor_finalize;
 }
 
-
 typedef enum _CcDisplayMonitorUnderscanning
 {
   UNDERSCANNING_UNSUPPORTED = 0,
@@ -325,6 +324,13 @@
   UNDERSCANNING_ENABLED
 } CcDisplayMonitorUnderscanning;
 
+typedef enum _CcDisplayMonitorVariableRefreshRate
+{
+  VRR_UNSUPPORTED = 0,
+  VRR_DISABLED,
+  VRR_ENABLED
+} CcDisplayMonitorVariableRefreshRate;
+
 struct _CcDisplayMonitorDBus
 {
   CcDisplayMonitor parent_instance;
@@ -340,6 +346,7 @@
   int height_mm;
   gboolean builtin;
   CcDisplayMonitorUnderscanning underscanning;
+  CcDisplayMonitorVariableRefreshRate vrr_mode;
   int max_width;
   int max_height;
 
@@ -640,6 +647,37 @@
     self->underscanning = UNDERSCANNING_DISABLED;
 }
 
+static gboolean
+cc_display_monitor_dbus_supports_variable_refresh_rate (CcDisplayMonitor *pself)
+{
+  CcDisplayMonitorDBus *self = CC_DISPLAY_MONITOR_DBUS (pself);
+
+  return self->vrr_mode != VRR_UNSUPPORTED;
+}
+
+static gboolean
+cc_display_monitor_dbus_get_variable_refresh_rate (CcDisplayMonitor *pself)
+{
+  CcDisplayMonitorDBus *self = CC_DISPLAY_MONITOR_DBUS (pself);
+
+  return self->vrr_mode == VRR_ENABLED;
+}
+
+static void
+cc_display_monitor_dbus_set_variable_refresh_rate (CcDisplayMonitor *pself,
+                                                   gboolean enabled)
+{
+  CcDisplayMonitorDBus *self = CC_DISPLAY_MONITOR_DBUS (pself);
+
+  if (self->vrr_mode == VRR_UNSUPPORTED)
+    return;
+
+  if (enabled)
+    self->vrr_mode = VRR_ENABLED;
+  else
+    self->vrr_mode = VRR_DISABLED;
+}
+
 static CcDisplayMode *
 cc_display_monitor_dbus_get_closest_mode (CcDisplayMonitorDBus *self,
                                           CcDisplayModeDBus *mode)
@@ -790,6 +828,7 @@
 cc_display_monitor_dbus_init (CcDisplayMonitorDBus *self)
 {
   self->underscanning = UNDERSCANNING_UNSUPPORTED;
+  self->vrr_mode = VRR_UNSUPPORTED;
   self->max_width = G_MAXINT;
   self->max_height = G_MAXINT;
 }
@@ -843,6 +882,9 @@
   parent_class->supports_underscanning = cc_display_monitor_dbus_supports_underscanning;
   parent_class->get_underscanning = cc_display_monitor_dbus_get_underscanning;
   parent_class->set_underscanning = cc_display_monitor_dbus_set_underscanning;
+  parent_class->supports_variable_refresh_rate = cc_display_monitor_dbus_supports_variable_refresh_rate;
+  parent_class->get_variable_refresh_rate = cc_display_monitor_dbus_get_variable_refresh_rate;
+  parent_class->set_variable_refresh_rate = cc_display_monitor_dbus_set_variable_refresh_rate;
   parent_class->set_mode = cc_display_monitor_dbus_set_mode;
   parent_class->set_compatible_clone_mode = cc_display_monitor_dbus_set_compatible_clone_mode;
   parent_class->set_position = cc_display_monitor_dbus_set_position;
@@ -920,6 +962,15 @@
           else
             self->underscanning = UNDERSCANNING_DISABLED;
         }
+      else if (g_str_equal (s, "is-vrr-enabled"))
+        {
+          gboolean vrr_enabled = FALSE;
+          g_variant_get (v, "b", &vrr_enabled);
+          if (vrr_enabled)
+            self->vrr_mode = VRR_ENABLED;
+          else
+            self->vrr_mode = VRR_DISABLED;
+        }
       else if (g_str_equal (s, "max-screen-size"))
         {
           g_variant_get (v, "ii", &self->max_width, &self->max_height);
@@ -1018,6 +1069,10 @@
                              "underscanning",
                              g_variant_new_boolean (monitor->underscanning == UNDERSCANNING_ENABLED));
 
+      g_variant_builder_add (&props_builder, "{sv}",
+                             "enable_vrr",
+                             g_variant_new_boolean (monitor->vrr_mode == VRR_ENABLED));
+
       mode_dbus = CC_DISPLAY_MODE_DBUS (monitor->current_mode);
       g_variant_builder_add (&builder, "(ss@*)",
                              monitor->connector_name,
@@ -1152,6 +1207,9 @@
       if (m1->underscanning != m2->underscanning)
         return FALSE;
 
+      if (m1->vrr_mode != m2->vrr_mode)
+        return FALSE;
+
       if (!cc_display_logical_monitor_equal (m1->logical_monitor, m2->logical_monitor))
         return FALSE;
 
--- a/panels/display/cc-display-config.h	2022-06-30 16:12:42.000000000 -0700
+++ b/panels/display/cc-display-config.h	2022-07-07 10:53:00.024034631 -0700
@@ -123,6 +123,10 @@
   gboolean          (*get_underscanning)      (CcDisplayMonitor  *self);
   void              (*set_underscanning)      (CcDisplayMonitor  *self,
                                                gboolean           u);
+  gboolean          (*supports_variable_refresh_rate) (CcDisplayMonitor *self);
+  gboolean          (*get_variable_refresh_rate)      (CcDisplayMonitor *self);
+  void              (*set_variable_refresh_rate)      (CcDisplayMonitor *self,
+                                                       gboolean          u);
   CcDisplayMode*    (*get_mode)               (CcDisplayMonitor  *self);
   CcDisplayMode*    (*get_preferred_mode)     (CcDisplayMonitor  *self);
   GList*            (*get_modes)              (CcDisplayMonitor  *self);
@@ -218,6 +222,11 @@
 void              cc_display_monitor_set_underscanning      (CcDisplayMonitor  *monitor,
                                                              gboolean           underscanning);
 
+gboolean          cc_display_monitor_supports_variable_refresh_rate (CcDisplayMonitor *monitor);
+gboolean          cc_display_monitor_get_variable_refresh_rate      (CcDisplayMonitor *monitor);
+void              cc_display_monitor_set_variable_refresh_rate      (CcDisplayMonitor *monitor,
+                                                                     gboolean          enabled);
+
 CcDisplayMode*    cc_display_monitor_get_mode               (CcDisplayMonitor  *monitor);
 void              cc_display_monitor_get_geometry           (CcDisplayMonitor  *monitor,
                                                              int               *x,
--- a/panels/display/cc-display-settings.c	2022-06-30 16:12:42.000000000 -0700
+++ b/panels/display/cc-display-settings.c	2022-07-07 10:54:24.748656319 -0700
@@ -57,6 +57,8 @@
   GtkWidget        *scale_combo_row;
   GtkWidget        *underscanning_row;
   GtkWidget        *underscanning_switch;
+  GtkWidget        *variable_refresh_rate_row;
+  GtkWidget        *variable_refresh_rate_switch;
 };
 
 typedef struct _CcDisplaySettings CcDisplaySettings;
@@ -255,6 +257,7 @@
       gtk_widget_set_visible (self->scale_combo_row, FALSE);
       gtk_widget_set_visible (self->scale_buttons_row, FALSE);
       gtk_widget_set_visible (self->underscanning_row, FALSE);
+      gtk_widget_set_visible (self->variable_refresh_rate_row, FALSE);
 
       return G_SOURCE_REMOVE;
     }
@@ -265,6 +268,7 @@
   g_object_freeze_notify ((GObject*) self->resolution_row);
   g_object_freeze_notify ((GObject*) self->scale_combo_row);
   g_object_freeze_notify ((GObject*) self->underscanning_switch);
+  g_object_freeze_notify ((GObject*) self->variable_refresh_rate_switch);
 
   cc_display_monitor_get_geometry (self->selected_output, NULL, NULL, &width, &height);
 
@@ -469,6 +473,12 @@
   gtk_switch_set_active (GTK_SWITCH (self->underscanning_switch),
                          cc_display_monitor_get_underscanning (self->selected_output));
 
+  gtk_widget_set_visible (self->variable_refresh_rate_row,
+                          cc_display_monitor_supports_variable_refresh_rate (self->selected_output) &&
+                          !cc_display_config_is_cloning (self->config));
+  gtk_switch_set_active (GTK_SWITCH (self->variable_refresh_rate_switch),
+                         cc_display_monitor_get_variable_refresh_rate (self->selected_output));
+
   self->updating = TRUE;
   g_object_thaw_notify ((GObject*) self->enabled_switch);
   g_object_thaw_notify ((GObject*) self->orientation_row);
@@ -476,6 +486,7 @@
   g_object_thaw_notify ((GObject*) self->resolution_row);
   g_object_thaw_notify ((GObject*) self->scale_combo_row);
   g_object_thaw_notify ((GObject*) self->underscanning_switch);
+  g_object_thaw_notify ((GObject*) self->variable_refresh_rate_switch);
   self->updating = FALSE;
 
   return G_SOURCE_REMOVE;
@@ -510,6 +521,21 @@
 }
 
 static void
+on_variable_refresh_rate_switch_active_changed_cb (GtkWidget         *widget,
+                                                   GParamSpec        *pspec,
+                                                   CcDisplaySettings *self)
+{
+  if (self->updating)
+    return;
+
+  cc_display_monitor_set_variable_refresh_rate (self->selected_output,
+                                                gtk_switch_get_active (GTK_SWITCH (self->variable_refresh_rate_switch)));
+
+  g_signal_emit_by_name (G_OBJECT (self), "updated", self->selected_output);
+}
+
+
+static void
 on_orientation_selection_changed_cb (GtkWidget         *widget,
                                      GParamSpec        *pspec,
                                      CcDisplaySettings *self)
@@ -760,6 +786,8 @@
   gtk_widget_class_bind_template_child (widget_class, CcDisplaySettings, scale_combo_row);
   gtk_widget_class_bind_template_child (widget_class, CcDisplaySettings, underscanning_row);
   gtk_widget_class_bind_template_child (widget_class, CcDisplaySettings, underscanning_switch);
+  gtk_widget_class_bind_template_child (widget_class, CcDisplaySettings, variable_refresh_rate_row);
+  gtk_widget_class_bind_template_child (widget_class, CcDisplaySettings, variable_refresh_rate_switch);
 
   gtk_widget_class_bind_template_callback (widget_class, on_enabled_switch_active_changed_cb);
   gtk_widget_class_bind_template_callback (widget_class, on_orientation_selection_changed_cb);
@@ -767,6 +795,7 @@
   gtk_widget_class_bind_template_callback (widget_class, on_resolution_selection_changed_cb);
   gtk_widget_class_bind_template_callback (widget_class, on_scale_selection_changed_cb);
   gtk_widget_class_bind_template_callback (widget_class, on_underscanning_switch_active_changed_cb);
+  gtk_widget_class_bind_template_callback (widget_class, on_variable_refresh_rate_switch_active_changed_cb);
 }
 
 static void
--- a/panels/display/cc-display-settings.ui	2022-06-30 16:12:42.000000000 -0700
+++ b/panels/display/cc-display-settings.ui	2022-07-07 10:54:46.322069248 -0700
@@ -56,6 +56,20 @@
           </object>
         </child>
         <child>
+          <object class="AdwActionRow" id="variable_refresh_rate_row">
+            <property name="width_request">100</property>
+            <property name="title" translatable="yes">Variable Refresh Rate</property>
+            <child>
+              <object class="GtkSwitch" id="variable_refresh_rate_switch">
+                <property name="can_focus">False</property>
+                <property name="halign">end</property>
+                <property name="valign">center</property>
+                <signal name="notify::active" handler="on_variable_refresh_rate_switch_active_changed_cb" swapped="no"/>
+              </object>
+            </child>
+          </object>
+        </child>
+        <child>
           <object class="AdwActionRow" id="underscanning_row">
             <property name="width_request">100</property>
             <property name="title" translatable="yes">Adjust for TV</property>
